src/index.ts

```diff
@@ -203,158 +232,217 @@ function getPropertyNames(schema: Schema) {
-      if (!prop.$ref && prop.type === 'object') {
-        return true;
-      }
+function objectHasBuffer(schema: Schema) {
+  const object = schema.xtpType as ObjectType
+  for (const prop of object.properties) {
+    if (prop.kind === 'buffer') {
+      return true
     }
   }
-  return false;
+  return false
+}
+
+function isPropertyRequired(property: Property) {
+  return property.required === undefined || property.required
 }
 
-function objectHasBuffer(object: Schema) {
+function numRequiredProperties(object: Schema) {
+  let count = 0;
   for (const prop of object.properties) {
-    if (!prop.$ref && prop.type === 'buffer') {
-      return true;
+    if (isPropertyRequired(prop)) {
+      count++
     }
   }
+  return count
 }
 
-function usesBuffer(objects: any) {
-  for (const object of objects) {
-    if (objectHasBuffer(object)) {
-      return true;
+function sortedProperties(object: Schema) {
+  const properties: Property[] = []
+  let curRequiredProperties = 0
+  for (const prop of object.properties) {
+    if (!isPropertyRequired(prop)) {
+      properties.push(prop)
+      continue;
     }
+    properties.splice(curRequiredProperties, 0, prop)
+    curRequiredProperties++
   }
-  return false;
+  return properties
 }
 
-function isBuffer(prop: Property) {
-  return (!prop.$ref && prop.type === 'buffer');
+function isTypeUntypedObject(type: XtpNormalizedType) {
+  if (type.kind === 'object') {
+    const object = type as ObjectType;
+    if (!object.properties?.length && object.name === '') {
+      return true
+    }
+  }
+  return false
 }
 
 export function render() {
   const tmpl = Host.inputString();
   const prevctx = getContext();
 
-  const enums: any[] = [];
-  const objects: any[] = [];
+  const enums: Schema[] = [];
+  const objects: Schema[] = [];
   Object.values(prevctx.schema.schemas).forEach(schema => {
-    if (schema.enum) {
+    if (helpers.isEnum(schema)) {
       enums.push(schema);
-    } else {
+    } else if (helpers.isObject(schema)) {
+      const object = schema.xtpType as ObjectType;
       // insertion sort objects ahead of objects that use them
       let i = 0;
       for (; i < objects.length; i++) {
-        if (objectReferencesObject(objects[i], schema.name)) {
+        if (objectReferencesObject(objects[i].xtpType as ObjectType, object.name)) {
           break;
         }
       }
-      objects.splice(i, 0, schema);
+      objects.splice(i, 0, schema); // we need Schema as it has the required attribute
+    } else {
+      throw new Error("unhandled schema type " + schema.xtpType.kind)
     }
   });
   // sort the properties for efficient struct layout
   for (const object of objects) {
     object.properties.sort((a: Property, b: Property) => {
-      return getEstimatedSize(b) - getEstimatedSize(a);
+      return v2GetEstimatedSize(b.xtpType) - v2GetEstimatedSize(a.xtpType)
     });
   }
 
+  const headerUsesJSON = function () {
+    for (const schema of objects) {
```

Comment:
Ah, I see – we don't need to recurse through the object definition tree because every object is expected to be available at top level in this `objects` list (we don't support nested definitions.)